src/Categories/Functor/Profunctor.agda

```diff
@@ -3,14 +3,399 @@
 module Categories.Functor.Profunctor where
 
 open import Level
+open import Data.Empty using (⊥)
+open import Data.Product using (_,_; _×_)
+open import Data.Sum using (_⊎_; inj₁; inj₂)
+open import Function using () renaming (_∘′_ to _∙_)
+open import Function.Equality using (_⟨$⟩_; cong)
+open import Relation.Binary using (Setoid; module Setoid)
+open import Relation.Binary.Construct.Closure.SymmetricTransitive as ST using (Plus⇔; minimal)
+import Relation.Binary.Reasoning.Setoid as SetoidR
 
 open import Categories.Category
-open import Categories.Category.Instance.Setoids
-open import Categories.Functor.Bifunctor
-open import Categories.Functor.Hom
+open import Categories.Category.Construction.Functors using (Functors; curry)
+open import Categories.Category.Instance.Cats using (Cats)
+open import Categories.Category.Instance.Setoids using (Setoids)
+open import Categories.Category.Product using ()
+open import Categories.Functor hiding (id)
+open import Categories.Functor.Bifunctor using (Bifunctor; appˡ; appʳ)
+open import Categories.Functor.Bifunctor.Properties using ([_]-commute)
+open import Categories.Functor.Construction.LiftSetoids using (LiftSetoids)
+open import Categories.Functor.Hom using (Hom[_][-,-])
+open import Categories.Morphism.Reasoning using (id-comm; id-comm-sym)
+open import Categories.NaturalTransformation using (NaturalTransformation; _∘ᵥ_; _∘ˡ_)
+open import Categories.NaturalTransformation.NaturalIsomorphism using (NaturalIsomorphism)
+open import Categories.NaturalTransformation.Properties using (appˡ-nat; appʳ-nat)
 
-Profunctor : ∀ {o ℓ e} {o′ ℓ′ e′} → Category o ℓ e → Category o′ ℓ′ e′ → Set _
-Profunctor {ℓ = ℓ} {e} {ℓ′ = ℓ′} {e′} C D = Bifunctor (Category.op D) C (Setoids (ℓ ⊔ ℓ′) (e ⊔ e′))
+open Setoid renaming (_≈_ to _[[_≈_]])
 
-id : ∀ {o ℓ e} → {C : Category o ℓ e} → Profunctor C C
-id {C = C} = Hom[ C ][-,-]
+record Profunctor {o ℓ e} {o′ ℓ′ e′} ℓ″ e″ (C : Category o ℓ e) (D : Category o′ ℓ′ e′) : Set (o ⊔ o′ ⊔ suc (ℓ ⊔ ℓ′ ⊔ ℓ″ ⊔ e ⊔ e′ ⊔ e″)) where
```

Comment:
The extra levels are necessary in general; e.g. you could have a profunctor from 1 to 1 with a proper class of heteromorphisms.  A common way to end up in this situation, at least theoretically, is composition, so I needed to add them.  Profunctors between small categories for a particular universe (all three levels the same) would be much cleaner with fewer lifts, but that doesn't fit the way the standard library or this library are set up.

Reply:
The 1-field record isn't *just* a place to hang helper functions and modules, it also aids inference.

Reply:
Agreed. If you've encountered issues with inference, then definitely this helps.